pkg/kafka/policy.go

```diff
@@ -0,0 +1,251 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/optiopay/kafka/proto"
+	log "github.com/sirupsen/logrus"
+)
+
+func produceTopicContained(neededTopic string, topics []proto.ProduceReqTopic) bool {
+	log.Debugf("Looking for topic %q in %+v", neededTopic, topics)
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchProduceReq(req *proto.ProduceReq, rule api.PortRuleKafka) bool {
+	if req == nil {
+		return false
+	}
+
+	if rule.Topic != "" && !produceTopicContained(rule.Topic, req.Topics) {
+		log.Debugf("denied (topic)")
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	log.Debugf("allowed produce")
+	return true
+}
+
+func fetchTopicContained(neededTopic string, topics []proto.FetchReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchFetchReq(req *proto.FetchReq, rule api.PortRuleKafka) bool {
+	if req == nil {
+		return false
+	}
+
+	if rule.Topic != "" && !fetchTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func offsetTopicContained(neededTopic string, topics []proto.OffsetReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetReq(req *proto.OffsetReq, rule api.PortRuleKafka) bool {
+	if req == nil {
+		return false
+	}
+
+	if rule.Topic != "" && !offsetTopicContained(rule.Topic, req.Topics) {
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	return true
+}
+
+func topicContained(neededTopic string, topics []string) bool {
+	for _, topic := range topics {
+		if topic == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchMetadataReq(req *proto.MetadataReq, rule api.PortRuleKafka) bool {
+	if req == nil {
+		return false
+	}
+
+	if rule.Topic != "" && !topicContained(rule.Topic, req.Topics) {
+		log.Debugf("Denied (topic)")
+		return false
+	}
+
+	if rule.ClientID != "" && rule.ClientID != req.ClientID {
+		return false
+	}
+
+	log.Debugf("allowed metadata")
+	return true
+}
+
+func offsetCommitTopicContained(neededTopic string, topics []proto.OffsetCommitReqTopic) bool {
+	for _, topic := range topics {
+		if topic.Name == neededTopic {
+			return true
+		}
+	}
+
+	return false
+}
+
+func matchOffsetCommitReq(req *proto.OffsetCommitReq, rule api.PortRuleKafka) bool {
```

Comment:
Nit (superficial): matchOffsetCommitReq matchOffsetReq seems to be duplicated code, other than the topic daatstructure type. Can be re-organized to avoid the duplicate code.

Reply:
I agree that this should be re-organized later. For now, the types are different. We need to change the parser types to provide an interface for the common fields. Definitely doable and desirable but not super critical to get it working as a first step.